pkg/operator/reports_test.go

```diff
@@ -92,3 +101,217 @@ func TestGetNextReportPeriod(t *testing.T) {
+			runningCond := isReportFinished(mockLogger, testCase.report)
+			assert.Equalf(t, runningCond, testCase.expectFinished, "expected the report '%s' would return '%t', but got '%t'", testCase.name, testCase.expectFinished, runningCond)
+		})
+	}
+}
+
+func TestValidateReport(t *testing.T) {
+	const (
+		testNamespace           = "default"
+		testReportName          = "test-report"
+		testQueryName           = "test-query"
+		testGenerationQueryName = "test-GQN"
+	)
+
+	ds1 := testhelpers.NewReportDataSource("datasource1", testNamespace)
+
+	testQuery := &metering.ReportGenerationQuery{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "test-query",
+			Namespace: testNamespace,
+		},
+		Spec: metering.ReportGenerationQuerySpec{
+			DataSources: []string{
+				"datasource1",
+			},
+		},
+	}
+
+	dataSourceGetter := testhelpers.NewReportDataSourceStore([]*metering.ReportDataSource{ds1})
+	queryGetter := testhelpers.NewReportGenerationQueryStore([]*metering.ReportGenerationQuery{testQuery})
+	reportGetter := testhelpers.NewReportStore(nil)
+
+	reportStart := &time.Time{}
+	reportEndTmp := reportStart.AddDate(0, 1, 0)
+	reportEnd := &reportEndTmp
+
+	testTable := []struct {
+		name         string
+		report       *metering.Report
+		queryName    string
+		expectErr    bool
+		expectErrMsg string
+	}{
+		{
+			name:         "empty spec.generationQuery returns err",
+			report:       testhelpers.NewReport(testReportName, testNamespace, "", reportStart, reportEnd, v1alpha1.ReportStatus{}, nil, false),
+			queryName:    "",
+			expectErr:    true,
+			expectErrMsg: "must set spec.generationQuery",
+		},
+		{
+			name:         "spec.ReportingStart > spec.ReportingEnd returns err",
+			report:       testhelpers.NewReport(testReportName, testNamespace, testQueryName, reportEnd, reportStart, v1alpha1.ReportStatus{}, nil, false),
+			queryName:    testGenerationQueryName,
+			expectErr:    true,
+			expectErrMsg: fmt.Sprintf("spec.reportingEnd (%s) must be after spec.reportingStart (%s)", reportStart.String(), reportEnd.String()),
+		},
+		{
+			name:         "spec.ReportingEnd is unset and spec.RunImmediately is set returns err",
+			report:       testhelpers.NewReport(testReportName, testNamespace, testQueryName, reportStart, nil, v1alpha1.ReportStatus{}, nil, true),
+			queryName:    testGenerationQueryName,
+			expectErr:    true,
+			expectErrMsg: "spec.reportingEnd must be set if report.spec.runImmediately is true",
+		},
+		{
+			name:         "spec.GenerationQueryName does not exist returns err",
+			report:       testhelpers.NewReport(testReportName, testNamespace, testQueryName, reportStart, reportEnd, v1alpha1.ReportStatus{}, nil, false),
+			queryName:    testGenerationQueryName,
+			expectErr:    true,
+			expectErrMsg: fmt.Sprintf("ReportGenerationQuery %s does not exist", testGenerationQueryName),
+		},
+	}
+
+	for _, testCase := range testTable {
+		testCase.report.Spec.GenerationQueryName = testCase.queryName
```

Comment:
I dislike this. We should instead just set the queryName we intend to use directly in each test case.